### ISSUE TITLE
mkYesod and HandlerSite with polymorphic type parameters

### ISSUE THREAD

**Issue by Daniel-Diaz** [yesodweb/yesod#1051]
Currently, it seems impossible to use ``myYesod`` with a type with at least one polymorphic type parameter. Something like this:

```haskell
data Foo a = Foo a

mkYesod "Foo" $(parseRoutesFile "foo-routes-file")
```

is not allowed.

```
   The first argument of ‘Yesod.Core.Types.HandlerT’
      should have kind ‘*’,
      but ‘Foo’ has kind ‘* -> *’
```

This means we are currently restricted to kind ``*``.

Is there anyway I can use a type with polymorphic variables? If not, how feasible is to add this feature? Easy, hard, time consuming?

Thanks.

**Comment by mgsloan** [comment]
Looks to me like this would be easy to add.  In `mkYesodGeneral`, [here](https://github.com/yesodweb/yesod/blob/b5077abdd988e267a1cdc7dcaff43674a64d1755/yesod-core/Yesod/Core/Internal/TH.hs#L67), you'd use `lookupTypeName` and `reify` to get info about the datatype.  (it's minorly unfortunate that the `mkYesod` functions don't take a `Name`..)

You'd then need to write an `Info -> Maybe Int` function that gets the arity of the datatype.  You'd then create new polymorphic variable names, and supply them in the`site`  type constructed  [here](https://github.com/yesodweb/yesod/blob/b5077abdd988e267a1cdc7dcaff43674a64d1755/yesod-core/Yesod/Core/Internal/TH.hs#L86).

**Comment by gregwebs** [comment]
We could create a `makeYesod` version that takes a `Name` and eventually deprecate the existing form.

**Comment by gregwebs** [comment]
Why do you need to parameterize `Foo`?

**Comment by andrewthad** [comment]
@gregwebs That's was I was wanting to know.

**Comment by MaxGabriel** [comment]
I believe this is closed by #1055
